pageserver/src/pgdatadir_mapping.rs

```diff
@@ -1330,58 +1350,68 @@ impl Timeline {
     }
 
     /// Get cached size of relation if it not updated after specified LSN
-    pub fn get_cached_rel_size(&self, tag: &RelTag, lsn: Lsn) -> Option<BlockNumber> {
-        let rel_size_cache = self.rel_size_cache.read().unwrap();
-        if let Some((cached_lsn, nblocks)) = rel_size_cache.map.get(tag) {
+    pub fn get_cached_rel_size(&self, tag: &RelTag, version: Version<'_>) -> Option<BlockNumber> {
+        let lsn = version.get_lsn();
+        let rel_size_cache = self.rel_size_latest_cache.read().unwrap();
+        if let Some((cached_lsn, nblocks)) = rel_size_cache.get(tag) {
```

Comment:
I asked somehwere else too. Should we make primary not query the snapshot cache? It's a waste of cycles.

Reply:
I also answered somewhere:)
Primary still can use value from snapshot cache. Assume situation when PS is restarted and so both caches are empty. Then some replica is started at `last_record_lsn` and get relation size at this LSN. Then primary node is connected. It's `effective_lsn` will be the same as `last_record_lsn`. So it can find cache value in snapshot cache instead of loading it from KV storage.

Reply:
Sure, I'm not arguing that the primary can never get a hit in the snapshot cache, but it's very unlikely.
I'd just keep things simple and not query it on the primary since I expect the hit rate to be bad.

Not a blocker though, so leaving this up to you. Feel free to resolve if you decide not to change it.

Reply:
I agree with you that it is very unlikely. But I still think that it is better not to exclude it.
First of all because I think that cache lookup overhead is negligible here (comparing with access to KV storage).
Also probability that primary relation size is not cached is in any case very small, so we actually very unlikely that we do not find entry in latest cache and have to perform lookup in snapshot cache.
Finally I want to make code a little bit simpler and "symmetric" (if we perform lookup in latest cache for replica, why we do not want to perform lookup in snapshot cache for master:)